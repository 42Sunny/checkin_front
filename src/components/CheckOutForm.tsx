import React, { useEffect, useRef, useState } from "react";
import classes from "../styles/components/CheckOutUi.module.css";
import useUser from "../utils/hooks/useUser";
import SlideButton from "./SlideButton";

interface CardNumberProps {
  cardNum: string;
}

const CardNumber: React.FC<CardNumberProps> = ({ cardNum }) => (
  <>
    <p>카드 번호</p>
    <p className={classes.cardNum}>{cardNum}</p>
  </>
);

interface IProps {
  handleCheckOut: (e: React.FormEvent<HTMLFormElement>) => Promise<void>;
}
const CheckOutForm: React.FC<IProps> = ({ handleCheckOut }) => {
  const [sliderValue, setSliderValue] = useState(0);
  const ref = useRef<HTMLFormElement>(null);
  const {
    user: { cardNum },
  } = useUser();

  useEffect(() => {
    if (sliderValue === 100) ref.current?.requestSubmit();
  }, [sliderValue]);

  return (
    <form ref={ref} onSubmit={handleCheckOut} className={classes["checkout-info-wrapper"]}>
      <CardNumber cardNum={cardNum} />
      <SlideButton value={sliderValue} setValue={setSliderValue} />
    </form>
  );
};

export default CheckOutForm;
