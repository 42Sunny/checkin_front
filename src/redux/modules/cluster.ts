import { createAction, ActionType } from "typesafe-actions";
import { formatOfficeHours } from "../../utils/time";

// actions
const SET_CLUSTER = "cluster/SET_CLUSTER";
const SET_CURRENT_USER_COUNT = "cluster/SET_CURRENT_USER_COUNT";
const SET_OFFICE_HOUR = "cluster/SET_OFFICE_HOUR";
const SET_LUNCH_TIME = "cluster/SET_LUNCH_TIME";
// action creators

export const setCluster =
  createAction(SET_CLUSTER)<Omit<Cluster, "officeHour" | "officeLunchTime">>();
export const setCurrentUserCount =
  createAction(SET_CURRENT_USER_COUNT)<{ gaepo: number; seocho: number }>();
export const setOfficeHour = createAction(SET_OFFICE_HOUR)<{ officeHour: string }>();
export const setLunchTime = createAction(SET_LUNCH_TIME)<{ officeLunchTime: string }>();
// type
const actions = { setCluster, setCurrentUserCount, setOfficeHour, setLunchTime };
type ClusterActions = ActionType<typeof actions>;

// initialState
const initialState: Cluster = {
  openAt: "",
  closeAt: "",
  seocho: 0,
  gaepo: 0,
  seochoLimitation: 0,
  gaepoLimitation: 0,
  officeHour: "",
  officeLunchTime: "",
};

// reducer
const cluster = (state = initialState, action: ClusterActions): Cluster => {
  switch (action.type) {
    case SET_CLUSTER: {
      const { openAt, closeAt, gaepo, seocho, gaepoLimitation, seochoLimitation } = action.payload;
      return {
        ...state,
        openAt,
        closeAt,
        gaepoLimitation,
        seochoLimitation,
        gaepo,
        seocho,
      };
    }
    case SET_CURRENT_USER_COUNT: {
      const { gaepo, seocho } = action.payload;
      return { ...state, gaepo, seocho };
    }
    case SET_OFFICE_HOUR: {
      const { officeHour } = action.payload;
      return { ...state, officeHour };
    }
    case SET_LUNCH_TIME: {
      const { officeLunchTime } = action.payload;
      return { ...state, officeLunchTime };
    }
    default:
      return state;
  }
};

export default cluster;
