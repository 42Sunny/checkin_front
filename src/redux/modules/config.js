// actions
const SET_CONFIG = 'SET_CONFIG';

// action creators
export const setConfig = data => {
  const { openAt, closeAt, seocho, gaepo } = data;
  return {
    type: SET_CONFIG,
    payload: {
      openAt: openAt,
      closeAt: closeAt,
      seocho: seocho,
      gaepo: gaepo
    }
  };
};

// initalState
const initalState = {
  openAt: null,
  closeAt: null,
  seocho: 0,
  gaepo: 0
};

// reducer
export const config = (state = initalState, action) => {
  switch (action.type) {
    case SET_CONFIG:
      return {
        ...state,
        openAt: action.payload.openAt,
        closeAt: action.payload.closeAt,
        seocho: action.payload.seocho,
        gaepo: action.payload.gaepo
      };
    default:
      return state;
  }
};
