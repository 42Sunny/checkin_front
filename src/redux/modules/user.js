import { DEFAULT_PROFILE } from '../../utils/utils';

// actions
const LOGIN = 'LOGIN';
const LOGOUT = 'LOGOUT';
const SET_USER = 'SET_USER';
const SET_CARDNUM = 'SET_CARDNUM';

// action creators
export const login = () => {
  return {
    type: LOGIN
  };
};

export const logout = () => {
  return {
    type: LOGOUT
  };
};

export const setUser = user => {
  return {
    type: SET_USER,
    payload: user
  };
};

export const setCardNum = cardNum => {
  return {
    type: SET_CARDNUM,
    payload: cardNum
  };
};

// initalState
const initalState = {
  isLogin: false,
  id: '',
  cardNum: '',
  status: 'out',
  profile: DEFAULT_PROFILE
};

// reducer
export const user = (state = initalState, action) => {
  switch (action.type) {
    case LOGIN:
      return {
        ...state,
        isLogin: true
      };
    case LOGOUT:
      return {
        ...state,
        isLogin: false
      };
    case SET_USER:
      return {
        ...state,
        id: action.payload.id,
        cardNum: action.payload.cardNum,
        status: action.payload.status,
        profile: action.payload.profile
      };
    case SET_CARDNUM:
      return {
        ...state,
        cardNum: action.payload
      };
    default:
      return state;
  }
};
