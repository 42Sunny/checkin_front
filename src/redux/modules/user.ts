import { createAction, ActionType } from "typesafe-actions";
import * as Sentry from "@sentry/react";
import DEFAULT_PROFILE from "../../assets/user-default.png";
// actions
const LOGIN = "user/LOGIN";
const LOGOUT = "user/LOGOUT";
const SET_USER = "user/SET_USER";
const SET_CARD_NUM = "user/SET_CARD_NUM";

// action creators
export const login = createAction(LOGIN)();
export const logout = createAction(LOGOUT)();
export const setUser = createAction(SET_USER)<Omit<User, "isLogin" | "isAdmin">>();
export const setCardNum = createAction(SET_CARD_NUM)<{ cardNum: string }>();

// type
const actions = { setCardNum, setUser, login, logout };
type UserActions = ActionType<typeof actions>;

// initialState
const initialState: User = {
  isLogin: false,
  id: "",
  cardNum: "",
  state: "checkOut",
  checkinAt: null,
  checkoutAt: null,
  profile: DEFAULT_PROFILE,
  isAdmin: false,
};

// reducer
const user = (state = initialState, action: UserActions) => {
  switch (action.type) {
    case LOGIN:
      return {
        ...state,
        isLogin: true,
      };
    case LOGOUT:
      return {
        ...state,
        isLogin: false,
      };

    case SET_USER: {
      const { cardNum, checkinAt, checkoutAt, id, profile, state: userState } = action.payload;
      Sentry.setUser({ username: id });
      return { ...state, id, cardNum, state: userState, checkinAt, checkoutAt, profile };
    }
    case SET_CARD_NUM:
      return {
        ...state,
        cardNum: action.payload.cardNum,
      };
    default:
      return state;
  }
};

export default user;
