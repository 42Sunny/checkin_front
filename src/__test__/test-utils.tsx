// test-utils.jsx
import { configureStore } from "@reduxjs/toolkit";
import { Queries, render, RenderResult } from "@testing-library/react";
import { createMemoryHistory } from "history";
import React, { ReactElement } from "react";
import { Provider } from "react-redux";
import { Router } from "react-router-dom";
// Import your own reducer
import renderer from "react-test-renderer";
import appStore from "../redux/configureStore";
import rootReducer, { RootState } from "../redux/modules";
import { initialState as clusterInitState } from "../redux/modules/cluster";
import { initialState as userInitState } from "../redux/modules/user";

const makeCustomStore = ({
  cluster,
  user,
}: {
  cluster?: typeof clusterInitState;
  user?: typeof userInitState;
} = {}) => {
  return {
    clusterReducer: { ...clusterInitState, ...cluster },
    userReducer: { ...userInitState, ...user },
  };
};

const customRender = (
  ui: ReactElement,
  preloadedState: RootState = makeCustomStore(),
  { store = configureStore({ reducer: rootReducer, preloadedState }), ...renderOptions } = {},
  history = createMemoryHistory(),
): RenderResult<Queries, HTMLElement> => {
  const Wrapper: any = ({ children }: { children: ReactElement }) => {
    return (
      <Provider store={store}>
        <Router history={history}>{children}</Router>
      </Provider>
    );
  };
  return render(ui, { wrapper: Wrapper, ...renderOptions });
};

const createWithWrapper = (children: ReactElement) =>
  renderer.create(
    <Provider store={appStore}>
      <Router history={createMemoryHistory()}>{children}</Router>
    </Provider>,
  );

// re-export everything
export * from "@testing-library/react";
// override render method
export { customRender as render, createWithWrapper, makeCustomStore };
