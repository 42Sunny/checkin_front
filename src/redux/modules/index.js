import { combineReducers } from 'redux';
import { user } from './user';
import { config } from './config';
import { status } from './status';

const rootReducer = combineReducers({
  user: user,
  config: config,
  status: status
});

export default rootReducer;
