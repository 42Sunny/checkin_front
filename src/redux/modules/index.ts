import { combineReducers } from "redux";
import userReducer from "./user";
import configReducer from "./config";
import statusReducer from "./status";

const rootReducer = combineReducers({
  userReducer,
  configReducer,
  statusReducer,
});

export type RootState = ReturnType<typeof rootReducer>;

export default rootReducer;
