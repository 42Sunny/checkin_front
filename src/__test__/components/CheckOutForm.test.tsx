import { fireEvent } from "@testing-library/react";
import CheckOutForm from "../../components/CheckOutForm";
import { initialState as userInitState } from "../../redux/modules/user";
import { createWithWrapper, makeCustomStore, render, screen } from "../test-utils";

describe("src/components/CheckOutForm.tsx", () => {
  let handleCheckOut: any;
  beforeEach(() => {
    handleCheckOut = jest.fn((e) => e.preventDefault());
  });
  it("matches snapshot", () => {
    const component = createWithWrapper(<CheckOutForm handleCheckOut={handleCheckOut} />);
    expect(component.toJSON()).toMatchSnapshot();
  });

  it("renders card number", () => {
    render(
      <CheckOutForm handleCheckOut={handleCheckOut} />,
      makeCustomStore({ user: { ...userInitState, cardNum: "1" } }),
    );
    const cardNumberEl = screen.queryByText("1");
    expect(cardNumberEl).toBeInTheDocument();
  });
  it("when user slide button to end handleSubmit gets called", () => {
    render(
      <CheckOutForm handleCheckOut={handleCheckOut} />,
      makeCustomStore({ user: { ...userInitState, cardNum: "1" } }),
    );
    const slideBtnEl = screen.getByRole("slider");

    fireEvent.input(slideBtnEl, { target: { value: "80" } });
    jest.resetAllMocks();
    fireEvent.touchEnd(slideBtnEl, { target: { value: "80" } });

    expect(handleCheckOut).toBeCalledTimes(1);
  });
});
