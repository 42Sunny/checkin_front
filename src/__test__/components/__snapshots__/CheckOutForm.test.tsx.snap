// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`src/components/CheckOutForm.tsx matches snapshot 1`] = `
<form
  className="checkout-info-wrapper"
>
  <p>
    카드 번호
  </p>
  <p
    className="cardNum"
  >
    
  </p>
  <div
    className="slider-wrapper"
  >
    <input
      className="slider"
      max={99}
      min={0}
      onInput={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      type="range"
      value={0}
    />
    <p
      className="slider-backgroundText"
      role="contentinfo"
    >
      밀어서 체크아웃
    </p>
  </div>
</form>
`;
